packages/remark-grid-tables/src/index.js

```diff
@@ -0,0 +1,469 @@
+          if (other._startPosition >= newCell._startPosition &&
+          other._endPosition <= newCell._endPosition) {
+            if (other._lines.length === 0) {
+              newCells.splice(ncc, 1)
+              ncc -= 1
+              if (nc > ncc) {
+                nc -= 1
+                newCell = newCells[nc]
+              }
+            }
+          }
+        }
+      }
+    }
+    this.lastRow()._cells = newCells
+  }
+
+  let TableRow = function (linesInfos) {
+    this._linesInfos = linesInfos
+    this._cells = []
+    for (let i = 0; i < linesInfos.length - 1; ++i) {
+      this._cells.push(new TableCell(linesInfos[i] + 1, linesInfos[i + 1]))
+    }
+  }
+
+  TableRow.prototype.updateContent = function (line) {
+    for (let c = 0; c < this._cells.length; ++c) {
+      const cell = this._cells[c]
+      cell._lines.push(line.substring(cell._startPosition, cell._endPosition))
+    }
+  }
+
+  let TableCell = function (startPosition, endPosition) {
+    this._startPosition = startPosition
+    this._endPosition = endPosition
+    this._colspan = 1
+    this._rowspan = 1
+    this._lines = []
+  }
+
+  TableCell.prototype.mergeWith = function (other) {
+    this._endPosition = other._endPosition
+    this._colspan += other._colspan
+    const newLines = []
+    for (let l = 0; l < this._lines.length; ++l) {
+      newLines.push(`${this._lines[l]}|${other._lines[l]}`)
+    }
+    this._lines = newLines
+  }
+
+  function extractTableContent (lines, linesInfos, hasHeader) {
+    const table = new Table(linesInfos)
+
+    for (let l = 0; l < lines.length; ++l) {
+      const line = lines[l]
+      const matchHeader = hasHeader & isHeaderLine(line) !== null
+      const isEndLine = matchHeader | isPartLine(line) !== null
+
+      if (isEndLine) {
+        // It is a header, a plain line or a line with plain line part.
+        // First, update the last row
+        table.lastPart().updateWithMainLine(line, isEndLine)
+
+        // Create the new raw
+        if (l !== 0) {
+          if (matchHeader) {
+            table.addPart()
+          } else if (isSeparationLine(line)) {
+            table.lastPart().addRow()
+          } else {
+            table.lastPart().updateWithPartLine(line)
+          }
+        }
+        // update the last row
+        table.lastPart().updateWithMainLine(line, isEndLine)
+      } else {
+        // it's a plain line
+        table.lastPart().updateWithMainLine(line, isEndLine)
+        table.lastPart().lastRow().updateContent(line)
+      }
+    }
+    // Because the last line is a separation, the last row is always empty
+    table.lastPart().removeLastRow()
+    return table
+  }
+
+  function generateTable (tableContent, now, tokenizer) {
+    let wrapperClassName = 'table-wrapper'
+    if (classNames.wrapper) {
+      wrapperClassName = classNames.wrapper
+    }
+    // Generate the gridTable node to insert in the AST
+    const tableWrapper = {
+      type: 'element',
```

Comment:
`element` is a [HAST](https://github.com/syntax-tree/hast) node, not a [MDAST](https://github.com/syntax-tree/mdast).

If we create `element` nodes, then this is a `rehype-` plugin -- it is working on HAST tree and not on MDAST tree.

`.data.hName` are HAST properties that can be defined on any HAST or MDAST nodes though, no need to change that.

I'm really in favour of having this plugin as `remark-` because we'll need these nodes in MDAST when we compile to LaTeX. I suggest the following: Since `Table` `TableCell` and `TableRow` exist in MDAST, use these node types instead of `element`. Get rid of the `tableWrapper`: it's rehype's job to wrap a table in a div if needed, the MD AST should know about this.

Keep all `node.data.h*` you currently have, I don't they'll get overridden by rehype. This way we would achieve:

* using MDAST node types
* having our nice `thead` and `rowspan` `colspan` etc